File: java/lib/src/test/java/io/zrz/httpsig/KeyPairTest.java
package io.zrz.httpsig;

import static org.junit.jupiter.api.Assertions.*;

import java.nio.file.Files;
import java.nio.file.Path;
import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.ArrayList;
import java.util.Base64;

import org.junit.jupiter.api.Test;

class KeyPairTest {

    private static final Path KEYS_DIR = Path.of("../../testdata/keys");

    // --- Auto-detection tests ---

    @Test
    void detectsRsaPss() throws Exception {
        PrivateKey priv = loadPkcs8PrivateKey(KEYS_DIR.resolve("rsa-pss.priv.pem"), "RSA");
        SigningKey sk = Keys.signingKey("rsa-test", priv);
        assertEquals(Algorithm.RSA_PSS_SHA512, sk.algorithm());
        assertEquals("rsa-test", sk.keyId());
    }

    @Test
    void detectsEcdsaP256() throws Exception {
        PrivateKey priv = loadEcPrivateKey(KEYS_DIR.resolve("ecc-p256.priv.pem"));
        SigningKey sk = Keys.signingKey("ec-test", priv);
        assertEquals(Algorithm.ECDSA_P256_SHA256, sk.algorithm());
    }

    @Test
    void detectsEd25519() throws Exception {
        PrivateKey priv = loadPkcs8PrivateKey(KEYS_DIR.resolve("ed25519.priv.pem"), "Ed25519");
        SigningKey sk = Keys.signingKey("ed-test", priv);
        assertEquals(Algorithm.ED25519, sk.algorithm());
    }

    // --- KeyPair round-trip tests ---

    @Test
    void rsaKeyPairRoundTrip() throws Exception {
        PrivateKey priv = loadPkcs8PrivateKey(KEYS_DIR.resolve("rsa-pss.priv.pem"), "RSA");
        PublicKey pub = loadSpkiPublicKey(KEYS_DIR.resolve("rsa-pss.pub.pem"), "RSA");
        io.zrz.httpsig.KeyPair kp = Keys.keyPair("rsa-kp", priv, pub);

        assertEquals("rsa-kp", kp.keyId());
        assertEquals(Algorithm.RSA_PSS_SHA512, kp.algorithm());

        byte[] data = "test data".getBytes();
        byte[] sig = kp.signingKey().sign(data);
        assertTrue(kp.verifyingKey().verify(data, sig));
    }

    @Test
    void ecdsaKeyPairRoundTrip() throws Exception {
        PrivateKey priv = loadEcPrivateKey(KEYS_DIR.resolve("ecc-p256.priv.pem"));
        PublicKey pub = loadSpkiPublicKey(KEYS_DIR.resolve("ecc-p256.pub.pem"), "EC");
        io.zrz.httpsig.KeyPair kp = Keys.keyPair("ec-kp", priv, pub);

        assertEquals(Algorithm.ECDSA_P256_SHA256, kp.algorithm());

        byte[] data = "test data".getBytes();
        byte[] sig = kp.signingKey().sign(data);
        assertTrue(kp.verifyingKey().verify(data, sig));
    }

    @Test
    void ed25519KeyPairRoundTrip() throws Exception {
        PrivateKey priv = loadPkcs8PrivateKey(KEYS_DIR.resolve("ed25519.priv.pem"), "Ed25519");
        PublicKey pub = loadSpkiPublicKey(KEYS_DIR.resolve("ed25519.pub.pem"), "Ed25519");
        io.zrz.httpsig.KeyPair kp = Keys.keyPair("ed-kp", priv, pub);

        assertEquals(Algorithm.ED25519, kp.algorithm());

        byte[] data = "test data".getBytes();
        byte[] sig = kp.signingKey().sign(data);
        assertTrue(kp.verifyingKey().verify(data, sig));
    }

    @Test
    void hmacKeyPairRoundTrip() throws Exception {
        byte[] secret = "super-secret-key-at-least-32-bytes!!".getBytes();
        io.zrz.httpsig.KeyPair kp = Keys.hmacKeyPair("hmac-kp", secret);

        assertEquals(Algorithm.HMAC_SHA256, kp.algorithm());
        assertEquals("hmac-kp", kp.keyId());

        byte[] data = "test data".getBytes();
        byte[] sig = kp.signingKey().sign(data);
        assertTrue(kp.verifyingKey().verify(data, sig));
    }

    @Test
    void jcaKeyPairBridge() throws Exception {
        var keyGen = java.security.KeyPairGenerator.getInstance("Ed25519");
        java.security.KeyPair jcaKp = keyGen.generateKeyPair();
        io.zrz.httpsig.KeyPair kp = Keys.keyPair("jca-ed", jcaKp);

        assertEquals(Algorithm.ED25519, kp.algorithm());

        byte[] data = "test data".getBytes();
        byte[] sig = kp.signingKey().sign(data);
        assertTrue(kp.verifyingKey().verify(data, sig));
    }

    @Test
    void keyPairRejectsIdMismatch() throws Exception {
        PrivateKey priv = loadPkcs8PrivateKey(KEYS_DIR.resolve("ed25519.priv.pem"), "Ed25519");
        PublicKey pub = loadSpkiPublicKey(KEYS_DIR.resolve("ed25519.pub.pem"), "Ed25519");
        SigningKey sk = Keys.ed25519SigningKey("a", priv);
        VerifyingKey vk = Keys.ed25519VerifyingKey("b", pub);
        assertThrows(IllegalArgumentException.class, () -> new io.zrz.httpsig.KeyPair(sk, vk));
    }

    @Test
    void keyPairRejectsAlgorithmMismatch() throws Exception {
        PrivateKey rsaPriv = loadPkcs8PrivateKey(KEYS_DIR.resolve("rsa-pss.priv.pem"), "RSA");
        PublicKey edPub = loadSpkiPublicKey(KEYS_DIR.resolve("ed25519.pub.pem"), "Ed25519");
        SigningKey sk = Keys.rsaPSSSigningKey("same-id", rsaPriv);
        VerifyingKey vk = Keys.ed25519VerifyingKey("same-id", edPub);
        assertThrows(IllegalArgumentException.class, () -> new io.zrz.httpsig.KeyPair(sk, vk));
    }

    @Test
    void keyPairRejectsNullSigningKey() {
        assertThrows(NullPointerException.class, () ->
            new io.zrz.httpsig.KeyPair(null, Keys.ed25519VerifyingKey("x",
                loadSpkiPublicKey(KEYS_DIR.resolve("ed25519.pub.pem"), "Ed25519"))));
    }

    @Test
    void autoDetectRejectsUnsupportedAlgorithm() {
        // DSA keys should be rejected
        assertThrows(IllegalArgumentException.class, () -> {
            var keyGen = java.security.KeyPairGenerator.getInstance("DSA");
            keyGen.initialize(2048);
            java.security.KeyPair dsa = keyGen.generateKeyPair();
            Keys.signingKey("dsa", dsa.getPrivate());
        });
    }

    // ---- Key loading helpers (same patterns as VectorTest) ----

    private static PrivateKey loadPkcs8PrivateKey(Path path, String algorithm) throws Exception {
        byte[] der = decodePem(Files.readString(path));
        var spec = new PKCS8EncodedKeySpec(der);
        try {
            return KeyFactory.getInstance(algorithm).generatePrivate(spec);
        } catch (java.security.spec.InvalidKeySpecException e) {
            if ("RSA".equals(algorithm)) {
                return KeyFactory.getInstance("RSASSA-PSS").generatePrivate(spec);
            }
            throw e;
        }
    }

    private static PrivateKey loadEcPrivateKey(Path path) throws Exception {
        String pem = Files.readString(path);
        if (pem.contains("BEGIN EC PRIVATE KEY")) {
            byte[] sec1Der = decodePem(pem);
            byte[] pkcs8Der = wrapEc256Sec1ToPkcs8(sec1Der);
            return KeyFactory.getInstance("EC").generatePrivate(new PKCS8EncodedKeySpec(pkcs8Der));
        }
        return loadPkcs8PrivateKey(path, "EC");
    }

    private static PublicKey loadSpkiPublicKey(Path path, String algorithm) throws Exception {
        byte[] der = decodePem(Files.readString(path));
        return KeyFactory.getInstance(algorithm).generatePublic(new X509EncodedKeySpec(der));
    }

    private static byte[] decodePem(String pem) {
        String b64 = pem.replaceAll("-----[A-Z ]+-----", "")
                        .replaceAll("\\s+", "");
        return Base64.getDecoder().decode(b64);
    }

    /**
     * Wraps a SEC1 EC private key in a PKCS#8 envelope for P-256.
     * Lifted from VectorTest to handle variable-length DER encoding correctly.
     */
    private static byte[] wrapEc256Sec1ToPkcs8(byte[] sec1Der) {
        int octetStringContentLen = sec1Der.length;
        int octetStringHeaderLen = octetStringContentLen < 128 ? 1 : 2;
        int totalSeqContent = 3 + 21 + 1 + octetStringHeaderLen + octetStringContentLen;

        var result = new ArrayList<Byte>();

        // outer SEQUENCE
        result.add((byte) 0x30);
        if (totalSeqContent < 128) {
            result.add((byte) totalSeqContent);
        } else {
            result.add((byte) 0x81);
            result.add((byte) totalSeqContent);
        }

        // version INTEGER 0
        result.add((byte) 0x02); result.add((byte) 0x01); result.add((byte) 0x00);

        // AlgorithmIdentifier SEQUENCE (21 bytes)
        result.add((byte) 0x30); result.add((byte) 0x13);
        // ecPublicKey OID
        for (byte b : new byte[] {0x06, 0x07, 0x2a, (byte)0x86, 0x48, (byte)0xce, 0x3d, 0x02, 0x01}) {
            result.add(b);
        }
        // prime256v1 OID
        for (byte b : new byte[] {0x06, 0x08, 0x2a, (byte)0x86, 0x48, (byte)0xce, 0x3d, 0x03, 0x01, 0x07}) {
            result.add(b);
        }

        // OCTET STRING containing SEC1 key
        result.add((byte) 0x04);
        if (octetStringContentLen < 128) {
            result.add((byte) octetStringContentLen);
        } else {
            result.add((byte) 0x81);
            result.add((byte) octetStringContentLen);
        }
        for (byte b : sec1Der) {
            result.add(b);
        }

        byte[] out = new byte[result.size()];
        for (int i = 0; i < result.size(); i++) {
            out[i] = result.get(i);
        }
        return out;
    }
}

File: kotlin/lib/src/test/kotlin/io/zrz/httpsig/KeyPairTest.kt
package io.zrz.httpsig

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.io.ByteArrayOutputStream
import java.nio.file.Files
import java.nio.file.Path
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.PrivateKey
import java.security.PublicKey
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import java.util.Base64

class KeyPairTest {

    private val keysDir = Path.of("../../testdata/keys")

    // --- Auto-detection ---

    @Test
    fun `detects RSA-PSS from private key`() {
        val priv = loadPkcs8PrivateKey("rsa-pss.priv.pem", "RSA")
        val sk = Keys.signingKey("rsa-test", priv)
        assertEquals(Algorithm.RsaPssSha512, sk.algorithm)
        assertEquals("rsa-test", sk.keyId)
    }

    @Test
    fun `detects ECDSA P-256 from private key`() {
        val priv = loadEcPrivateKey("ecc-p256.priv.pem")
        val sk = Keys.signingKey("ec-test", priv)
        assertEquals(Algorithm.EcdsaP256Sha256, sk.algorithm)
    }

    @Test
    fun `detects Ed25519 from private key`() {
        val priv = loadPkcs8PrivateKey("ed25519.priv.pem", "Ed25519")
        val sk = Keys.signingKey("ed-test", priv)
        assertEquals(Algorithm.Ed25519, sk.algorithm)
    }

    @Test
    fun `detects RSA from public key`() {
        val pub = loadSpkiPublicKey("rsa-pss.pub.pem", "RSA")
        val vk = Keys.verifyingKey("rsa-test", pub)
        assertEquals(Algorithm.RsaPssSha512, vk.algorithm)
    }

    @Test
    fun `detects ECDSA from public key`() {
        val pub = loadSpkiPublicKey("ecc-p256.pub.pem", "EC")
        val vk = Keys.verifyingKey("ec-test", pub)
        assertEquals(Algorithm.EcdsaP256Sha256, vk.algorithm)
    }

    @Test
    fun `detects Ed25519 from public key`() {
        val pub = loadSpkiPublicKey("ed25519.pub.pem", "Ed25519")
        val vk = Keys.verifyingKey("ed-test", pub)
        assertEquals(Algorithm.Ed25519, vk.algorithm)
    }

    // --- KeyPair round-trips ---

    @Test
    fun `RSA key pair round-trip`() {
        val priv = loadPkcs8PrivateKey("rsa-pss.priv.pem", "RSA")
        val pub = loadSpkiPublicKey("rsa-pss.pub.pem", "RSA")
        val kp = Keys.keyPair("rsa-kp", priv, pub)

        assertEquals("rsa-kp", kp.keyId)
        assertEquals(Algorithm.RsaPssSha512, kp.algorithm)

        val data = "test data".toByteArray()
        val sig = kp.signingKey.sign(data)
        assertTrue(kp.verifyingKey.verify(data, sig))
    }

    @Test
    fun `ECDSA key pair round-trip`() {
        val priv = loadEcPrivateKey("ecc-p256.priv.pem")
        val pub = loadSpkiPublicKey("ecc-p256.pub.pem", "EC")
        val kp = Keys.keyPair("ec-kp", priv, pub)

        assertEquals(Algorithm.EcdsaP256Sha256, kp.algorithm)

        val data = "test data".toByteArray()
        val sig = kp.signingKey.sign(data)
        assertTrue(kp.verifyingKey.verify(data, sig))
    }

    @Test
    fun `Ed25519 key pair round-trip`() {
        val priv = loadPkcs8PrivateKey("ed25519.priv.pem", "Ed25519")
        val pub = loadSpkiPublicKey("ed25519.pub.pem", "Ed25519")
        val kp = Keys.keyPair("ed-kp", priv, pub)

        assertEquals(Algorithm.Ed25519, kp.algorithm)

        val data = "test data".toByteArray()
        val sig = kp.signingKey.sign(data)
        assertTrue(kp.verifyingKey.verify(data, sig))
    }

    @Test
    fun `HMAC key pair round-trip`() {
        val secret = "super-secret-key-at-least-32-bytes!!".toByteArray()
        val kp = Keys.hmacKeyPair("hmac-kp", secret)

        assertEquals(Algorithm.HmacSha256, kp.algorithm)
        assertEquals("hmac-kp", kp.keyId)

        val data = "test data".toByteArray()
        val sig = kp.signingKey.sign(data)
        assertTrue(kp.verifyingKey.verify(data, sig))
    }

    @Test
    fun `JCA key pair bridge`() {
        val keyGen = KeyPairGenerator.getInstance("Ed25519")
        val jcaKp = keyGen.generateKeyPair()
        val kp = Keys.keyPair("jca-ed", jcaKp)

        assertEquals(Algorithm.Ed25519, kp.algorithm)

        val data = "test data".toByteArray()
        val sig = kp.signingKey.sign(data)
        assertTrue(kp.verifyingKey.verify(data, sig))
    }

    // --- Validation ---

    @Test
    fun `rejects mismatched key IDs`() {
        val priv = loadPkcs8PrivateKey("ed25519.priv.pem", "Ed25519")
        val pub = loadSpkiPublicKey("ed25519.pub.pem", "Ed25519")
        val sk = Keys.ed25519SigningKey("a", priv)
        val vk = Keys.ed25519VerifyingKey("b", pub)
        assertThrows<IllegalArgumentException> { KeyPair(sk, vk) }
    }

    @Test
    fun `rejects mismatched algorithms`() {
        val rsaPriv = loadPkcs8PrivateKey("rsa-pss.priv.pem", "RSA")
        val edPub = loadSpkiPublicKey("ed25519.pub.pem", "Ed25519")
        val sk = Keys.rsaPSSSigningKey("x", rsaPriv)
        val vk = Keys.ed25519VerifyingKey("x", edPub)
        assertThrows<IllegalArgumentException> { KeyPair(sk, vk) }
    }

    // --- Key loading helpers (same approach as VectorTest) ---

    private fun loadPkcs8PrivateKey(filename: String, algorithm: String): PrivateKey {
        val der = decodePem(Files.readString(keysDir.resolve(filename)))
        val spec = PKCS8EncodedKeySpec(der)
        return try {
            KeyFactory.getInstance(algorithm).generatePrivate(spec)
        } catch (_: java.security.spec.InvalidKeySpecException) {
            if (algorithm == "RSA") {
                KeyFactory.getInstance("RSASSA-PSS").generatePrivate(spec)
            } else {
                throw IllegalStateException("cannot load private key: $filename as $algorithm")
            }
        }
    }

    private fun loadSpkiPublicKey(filename: String, algorithm: String): PublicKey {
        val der = decodePem(Files.readString(keysDir.resolve(filename)))
        val spec = X509EncodedKeySpec(der)
        return KeyFactory.getInstance(algorithm).generatePublic(spec)
    }

    private fun loadEcPrivateKey(filename: String): PrivateKey {
        val pem = Files.readString(keysDir.resolve(filename))
        if (pem.contains("BEGIN EC PRIVATE KEY")) {
            val sec1Der = decodePem(pem)
            val pkcs8Der = wrapEc256Sec1ToPkcs8(sec1Der)
            return KeyFactory.getInstance("EC").generatePrivate(PKCS8EncodedKeySpec(pkcs8Der))
        }
        return loadPkcs8PrivateKey(filename, "EC")
    }

    /**
     * Wraps a SEC1 EC private key in a PKCS#8 envelope for P-256.
     * Exactly matches the approach in VectorTest.
     */
    private fun wrapEc256Sec1ToPkcs8(sec1Der: ByteArray): ByteArray {
        val out = ByteArrayOutputStream()

        val version = byteArrayOf(0x02, 0x01, 0x00)

        val algId = byteArrayOf(
            0x30, 0x13,
            0x06, 0x07, 0x2a, 0x86.toByte(), 0x48, 0xce.toByte(), 0x3d, 0x02, 0x01,
            0x06, 0x08, 0x2a, 0x86.toByte(), 0x48, 0xce.toByte(), 0x3d, 0x03, 0x01, 0x07,
        )

        val octetStringHeader = if (sec1Der.size < 128) {
            byteArrayOf(0x04, sec1Der.size.toByte())
        } else {
            byteArrayOf(0x04, 0x81.toByte(), sec1Der.size.toByte())
        }

        val totalSeqContent = version.size + algId.size + octetStringHeader.size + sec1Der.size

        if (totalSeqContent < 128) {
            out.write(byteArrayOf(0x30, totalSeqContent.toByte()))
        } else {
            out.write(byteArrayOf(0x30, 0x81.toByte(), totalSeqContent.toByte()))
        }

        out.write(version)
        out.write(algId)
        out.write(octetStringHeader)
        out.write(sec1Der)

        return out.toByteArray()
    }

    private fun decodePem(pem: String): ByteArray {
        val b64 = pem.replace(Regex("-----[A-Z ]+-----"), "")
            .replace(Regex("\\s+"), "")
        return Base64.getDecoder().decode(b64)
    }
}
